ext/dom/tests/modern/extensions/Element_getInScopeNamespaces.phpt

```diff
@@ -0,0 +1,56 @@
+--TEST--
+DOM\Element::getInScopeNamespaces()
+--EXTENSIONS--
+dom
+--FILE--
+<?php
+
+$dom = DOM\XMLDocument::createFromString(<<<XML
+<root xmlns="urn:a">
+    <child xmlns="">
+        <c:child xmlns:c="urn:c"/>
+    </child>
+    <b:sibling xmlns:b="urn:b" xmlns:d="urn:d" d:foo="bar">
+        <d:child xmlns:d="urn:d2"/>
+    </b:sibling>
+</root>
+XML);
+
+var_dump($dom->getElementsByTagName('c:child')[0]->getInScopeNamespaces());
+var_dump($dom->getElementsByTagName('child')[0]->getInScopeNamespaces());
+var_dump($dom->getElementsByTagName('b:sibling')[0]->getInScopeNamespaces());
+var_dump($dom->getElementsByTagName('d:child')[0]->getInScopeNamespaces());
+var_dump($dom->documentElement->getInScopeNamespaces());
```

Comment:
I think technically this is correct since the element only knows about `xmlns="urn:a", but what I am looking for in `optimize_namespaces` is a way to list all namespaces declared inside the element and its children. (especially the xmlns attribute value (URI) since the prefix will be optimized out)

This probably comes with a complexity that a URI can have multiple localNames in different child elements.

Reply:
Right that makes sense, you also want the namespaces that were shadowed of course.
Then we'd need to return an array where each prefix corresponds to an array of possible uris.

Reply:
Right. So a getInScopeNamespaces and getDescendantNamespaces method. That should work.

If getDescendantNamespaces returns an array like getInScopeNamespaces does, then for an xml like this:
```
<parent xmlns:x="urn:foo">
  <child xmlns:x="urn:bar"/>
</parent>
```
It will return an array with key x, with value `["urn:foo", "urn:bar"]`.

In the array you just get the namespace uris, that's it. You don't get on which element the prefix+uri combination is in scope, which is a downgrade from old DOM.

It makes me think whether it makes sense to somehow reintroduce the NamespaceNode class anyway, just for xpath. The problem with this is that it's not really a node, old DOM just misused it to be like a node (instead of a Attr instance). Although I could call it NamespaceInfo or something like that. One of the consequences is that the return types of XPath::query and XPath::evaluate have to include NamespaceInfo, as it doesn't inherit from Node. This is a problem for static analysis tools like psalm, who refused to update the return type for those methods in old DOM because it would cause many warnings.

Reply:
The way I see it : If we have the methods you introduce above, we don't need to run `namespace::*` queries on xpath anymore. This would result in xpath only returning `Node`s and not some custom object. I'm not sure, but how do other languages cope with this `namespace`-axis?

Introducing a `NamespaceInfo` value object as return value might be better than the multi-dimensional array. The functions would just return `list<NamespaceInfo>` and the end-user can decide wether to group based on namespaceURI or local-name.


So my current feeling is to introduce the object, but not make it part of xpath.

Reply:
Your suggestion makes sense, and would avoid problems and confusion.

Browsers don't return anything on the namespace axis, and IIRC C# also doesn't. I'll have another look at others.

Reply:
Okay that took a long time, got held up with other stuff to do.
I've now added a NamespaceInfo class, and there's now both `Element::getInScopeNamespaces` and `Element::getDescendantNamespaces` which correspond to `./namespace::*` resp. `.//namespace::*`. At least I hope so :sweat_smile:

Reply:
No problem. It's not urgent 🙂

I'll play around with it in the coming week and let you know how my tests behave.

Reply:
It seems to be a bit off, see:

* https://github.com/veewee/xml/blob/5144ee213f47976063efce545ff7ae663acccd7f/tests/Xml/Dom/Locator/Xmlns/LinkedNamespacesTest.php
* https://github.com/veewee/xml/blob/5144ee213f47976063efce545ff7ae663acccd7f/tests/Xml/Dom/Locator/Xmlns/RecursiveLinkedNamespacesTest.php

The linked one (using `getInScopeNamespaces`) returns the same result as `getDescendantNamespaces` for the `item` here - which is not in line with `./namespace::*`.

```xml
<root xmlns="http://hello.com" xmlns:world="http://world.com" >
    <item>1</item>        
</root>
```

Not sure if it's not correct or that my understanding of it is wrong through :)

Reply:
Okay, let's first check the `LinkedNamespacesTest`.

`./namespace::*` and `.//namespace::*` return the same thing for `item`: https://3v4l.org/FcavJ
`getDescendantNamespaces` should correspond to `.//namespace::*` which means "inclusive descendant namespace axis".
In the output we see the same result for both XPath queries.

Comparing with the new methods, I get this output from the script below:
```
prefix: null
uri: http://hello.com
prefix: "world"
uri: http://world.com
---
prefix: null
uri: http://hello.com
prefix: "world"
uri: http://world.com
```

script:
```php
<?php

$xml = <<<XML
<root xmlns="http://hello.com" xmlns:world="http://world.com" >
    <item>1</item>        
</root>
XML;

$dom = DOM\XMLDocument::createFromString($xml);

foreach ($dom->documentElement->firstElementChild->getInScopeNamespaces() as $x) {
	echo "prefix: ", json_encode($x->prefix), "\n";
	echo "uri: ", $x->namespaceURI, "\n";
}

echo "---\n";

foreach ($dom->documentElement->firstElementChild->getDescendantNamespaces() as $x) {
	echo "prefix: ", json_encode($x->prefix), "\n";
	echo "uri: ", $x->namespaceURI, "\n";
}
```

So this seems to be right (except the xml namespace that XPath always adds).

I think what's going on is that your _existing_ test on branch v3.1.0 is wrong: https://github.com/veewee/xml/blob/47504aec3a1a8146ff7e870acce07b561e3c2947/tests/Xml/Dom/Locator/Xmlns/LinkedNamespacesTest.php#L41
In this line, you _think_ you get `<item>` by doing `$element->childNodes->item(0)` but that's not true. The first child node is the whitespace in front of `<item>`. If you replace that with `firstElementChild` you get the same behaviour as with the test on the spec-compliance branch.

Reply:
`RecursiveLinkedNamespacesTest` fails because of inconsistent namespace ordering. This happens because my code traverses the tree backwards: from bottom to top, and from attributes left to right.
I think the more logical approach is if the namespaces in the array appear in the ordering they appear as attributes in the document. I pushed a commit that fixes this and that fixes your test.

It's worth noting that libxml does list the namespaces from top to bottom but from right to left... I think the behaviour that I changed to is better.

Reply:
Allright cool! Thanks for pointing out the error in my previous test-case.
I got everything working in my lib now. Except there is one more issue left regarding namespace optimization. I'll try to figure it out later, it's most likely on my end.

Keep you posted!

Reply:
Cool, thanks for the effort!